### PR TITLE
Archives filters

### PR DESCRIPTION
**Pull request by funkyfuture**
follow-up of #1369

this adds the arguments `--first`, `--last` and `--sort-by`, and groups them with `--prefix` as archives filters that can be applied on repository targets.

usable with `check`, `delete`, `info` and `list`.

use-case in a script:

```bash
while used_space_exceeds_pcent; do
    log 'Used space exceeds ${MAX_DISK_USAGE}%, deleting oldest backup archive.'
    borg delete --first 1 --force $BORG_REPOSITORY & wait
done
```

### CODE REVIEW

**Comment by funkyfuture** [comment]
strange, locally 42cd9805a47198ee4e8cce9441808f10e9841dc1 appears as second of these commits.

File: src/borg/archiver.py [hunk 1]
```diff
@@ -2281,6 +2368,23 @@ def build_parser(self, prog=None):
...
+        sort_by_default = 'timestamp'
+        filters_group.add_argument('--sort-by', dest='sort_by', type=sort_by_spec, default=sort_by_default,
+                               help='Comma-separated list of sorting keys; valid keys are: {}; default is: {}'
+                               .format(', '.join(HUMAN_SORT_KEYS), sort_by_default))
+
+        group = filters_group.add_mutually_exclusive_group()
+        group.add_argument('--first', dest='first', metavar='N', default=0, type=int,
+                           help='delete N first archives')
+        group.add_argument('--last', dest='last', metavar='N', default=0, type=int,
+                           help='delete N last archives')
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: it always says "delete", no matter to where you add these filter args.
so, how about "consider N first/last archives".

File: src/borg/helpers.py [hunk 2]
```diff
@@ -143,10 +143,15 @@ def __delitem__(self, name):
         del self._archives[name]
 
     def list(self, sort_by=None, reverse=False):
-        # inexpensive Archive.list_archives replacement if we just need .name, .id, .ts
-        archives = self.values()  # [self[name] for name in self]
+        """ Inexpensive Archive.list_archives replacement if we just need .name, .id, .ts
+
+            :rtype: A :class:`list` of :class:`borg.helpers.ArchiveInfo` instances
+        """
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: please have a look at the usual docstring style and use the same. do a fixup commit.

File: src/borg/archiver.py [hunk 3]
```diff
@@ -2343,6 +2358,27 @@ def run(self, args):
             logger.warning("Using a pure-python msgpack! This will result in lower performance.")
         return args.func(args)
 
+    def _get_filtered_archives(self, args, manifest):
+        if args.location.archive:
+            logger.error('The options --first and --last can only be used on repository targets.')
+            self.exit_code = EXIT_ERROR
+            return []
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: Hmm, instead of logging an error and **continuing**, you could raise Error('...') which will terminate borg, with error status and show the given error msg.

File: src/borg/archiver.py [hunk 4]
```diff
@@ -2343,6 +2358,27 @@ def run(self, args):
...
+
+        archives = manifest.archives.list()
+        if not archives:
+            logger.critical('There are no archives.')
+            self.exit_code = self.exit_code or EXIT_WARNING
+            return []
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: same here.

File: src/borg/archiver.py [hunk 5]
```diff
@@ -2343,6 +2358,27 @@ def run(self, args):
...
 
+    def _get_filtered_archives(self, args, manifest):
+        if args.location.archive:
+            logger.error('The options --first and --last can only be used on repository targets.')
+            self.exit_code = EXIT_ERROR
+            return []
+
+        archives = manifest.archives.list()
+        if not archives:
+            logger.critical('There are no archives.')
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: why is this critical, why is this an error?

File: src/borg/archiver.py [hunk 6]
```diff
@@ -2343,6 +2358,27 @@ def run(self, args):
...
+            return []
+
+        for sortkey in reversed(args.sort_by.split(',')):
+            archives.sort(key=attrgetter(sortkey))
+        if args.last:
+            archives.reverse()
+
+        n = args.first or args.last
+
+        return archives[:n]
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: nice :)

File: src/borg/helpers.py (hunk 2 above)
Comment (enkore) [inline src/borg/helpers.py]: (Which is basically [PEP 0257](https://www.python.org/dev/peps/pep-0257/) as per PEP 8 - ie. textual, imperative style "Return that-n-that" etc., `:param:` stuff only for many params; no type annotations)

File: src/borg/archiver.py [hunk 7]
```diff
@@ -2343,6 +2358,27 @@ def run(self, args):
...
+            logger.critical('There are no archives.')
+            self.exit_code = self.exit_code or EXIT_WARNING
+            return []
+
+        for sortkey in reversed(args.sort_by.split(',')):
+            archives.sort(key=attrgetter(sortkey))
+        if args.last:
+            archives.reverse()
+
+        n = args.first or args.last
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: slightly better IMHO, but equivalent:
```
        if args.first:
            n = args.first
        else:  # if args.last
            n = args.last
            archives.reverse()
````

File: src/borg/archiver.py [hunk 8]
```diff
@@ -2343,6 +2358,27 @@ def run(self, args):
             logger.warning("Using a pure-python msgpack! This will result in lower performance.")
         return args.func(args)
 
+    def _get_filtered_archives(self, args, manifest):
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: nice method.

File: src/borg/helpers.py [hunk 9]
```diff
@@ -645,6 +650,16 @@ def replace_placeholders(text):
...
+def sort_by_spec(text):
+    for token in text.split(','):
+        if token not in HUMAN_SORT_KEYS:
+            raise ValueError('Invalid sort key: %s' % token)
+    return text.replace('timestamp', 'ts')
+
+
 def safe_timestamp(item_timestamp_ns):
     try:
         return datetime.fromtimestamp(bigint_to_int(item_timestamp_ns) / 1e9)
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: maybe do all the fixups into 1 fixup commit for this commit, so you can git rebase -i later and squash the fixup commit into the base commit.

File: src/borg/testsuite/archiver.py [hunk 10]
```diff
@@ -985,8 +988,13 @@ def test_delete(self):
         self.cmd('init', self.repository_location)
         self.cmd('create', self.repository_location + '::test', 'input')
         self.cmd('create', self.repository_location + '::test.2', 'input')
+        self.cmd('create', self.repository_location + '::test.3', 'input')
+        self.cmd('create', self.repository_location + '::another_test.1', 'input')
+        self.cmd('create', self.repository_location + '::another_test.2', 'input')
         self.cmd('extract', '--dry-run', self.repository_location + '::test')
         self.cmd('extract', '--dry-run', self.repository_location + '::test.2')
+        self.cmd('delete', '--prefix', 'another_', self.repository_location)
+        self.cmd('delete', '--last', '1', self.repository_location)
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: this is good for increasing coverage, but you have no assertion here, so it just tests that the code is not crashing.

File: src/borg/testsuite/archiver.py [hunk 11]
```diff
@@ -57,6 +57,9 @@ def exec_cmd(*args, archiver=None, fork=False, exe=None, **kw):
         except subprocess.CalledProcessError as e:
             output = e.output
             ret = e.returncode
+        except SystemExit as e:
+            output = ''
+            ret = e.code
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: how is that change related to this changeset?

File: src/borg/archiver.py [hunk 12]
```diff
@@ -30,7 +30,7 @@
 from .helpers import EXIT_SUCCESS, EXIT_WARNING, EXIT_ERROR
 from .helpers import Error, NoManifestError
 from .helpers import location_validator, archivename_validator, ChunkerParams, CompressionSpec
-from .helpers import PrefixSpec, sort_by_spec, HUMAN_SORT_KEYS
+from .helpers import prefix_spec, sort_by_spec, HUMAN_SORT_KEYS
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: hmm, we have also CompressionSpec (and ChunkerParams), so CamelCase has matched that.

File: src/borg/archiver.py [hunk 13]
```diff
@@ -2285,12 +2285,16 @@ def build_parser(self, prog=None):
 
     @staticmethod
     def add_archives_filters_args(subparser):
+        filters_group = subparser.add_argument_group('filters', 'Archive filters can be applied on repository targets.')
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: applied **to**?

File: src/borg/archiver.py [hunk 14]
```diff
@@ -2285,12 +2285,16 @@ def build_parser(self, prog=None):
...
+        filters_group.add_argument('-P', '--prefix', dest='prefix', type=prefix_spec, default='',
+                               help='only consider archive names starting with this prefix')
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: at the other places (old code), there is no default, why do you have one here?

please note that a empty string prefix matches all archives, which might be dangerous if the operation is "delete".

File: src/borg/archiver.py [hunk 15]
```diff
@@ -204,18 +204,55 @@ def do_check(self, args, repository):
...
+                if args.location.archive:
+                    archive_info = manifest.archives.get(args.location.archive)
+                    if archive_info is None:
+                        logger.error('Archive %s not found.' % args.location.archive)
+                        return EXIT_ERROR
+                    archives = (archive_info,)
+                elif any((args.first, args.last, args.prefix)):
+                    archives = self._get_filtered_archives(args, manifest)
+                else:
+                    archives = manifest.archives.list(prefix=args.prefix)
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: that looks strange.
232 will trigger if args.prefix is truish.
but in 235 you give args.prefix as a param (can it be anything else than None or '' here?).

File: src/borg/archiver.py (hunk 7 above)
Comment (funkyfuture) [inline src/borg/archiver.py]: but a use with `--prefix` only would then also require:

```
if not n:
    n = len(archives)
```

File: src/borg/testsuite/archiver.py (hunk 10 above)
Comment (funkyfuture) [inline src/borg/testsuite/archiver.py]: the assertion is that the command will succeed with exit code 0.

File: src/borg/testsuite/archiver.py (hunk 11 above)
Comment (funkyfuture) [inline src/borg/testsuite/archiver.py]: the tests for bad filters wouln't work as argparse raises `SystemExit` which would abort pytest.

File: src/borg/archiver.py (hunk 7 above)
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: that would make elif args.last: ... else: ...

File: src/borg/testsuite/archiver.py (hunk 10 above)
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: ah, ok.

File: src/borg/testsuite/archiver.py (hunk 11 above)
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: ok, maybe put just that into a src comment there?

Comment (enkore) [inline src/borg/testsuite/archiver.py]: Hmm, since that only applies to in-source (vs. binary) testing this change should rather go into [exec_cmd](https://github.com/borgbackup/borg/blob/master/src/borg/testsuite/archiver.py#L61)

File: src/borg/archiver.py (hunk 7 above)
Comment (funkyfuture) [inline src/borg/archiver.py]: sure, but uses seven lines instead of one. is simple not better than complex? imo the oneliner is also obvious.

File: src/borg/helpers.py (hunk 9 above)
Comment (funkyfuture) [inline src/borg/helpers.py]: what exactly are you refreing to here?

File: src/borg/testsuite/archiver.py (hunk 11 above)
Comment (funkyfuture) [inline src/borg/testsuite/archiver.py]: @enkore okay, but it is in that function.

File: src/borg/archiver.py (hunk 12 above)
Comment (funkyfuture) [inline src/borg/archiver.py]: no pep8 compliance?

File: src/borg/archiver.py (hunk 14 above)
Comment (funkyfuture) [inline src/borg/archiver.py]: with `args.prefix == None` [this list comprehension](https://github.com/borgbackup/borg/pull/1554/files#diff-d3eb25b2d47fdf880fd23aae789dd9efR150) would break.

no, it isn't dangerous to use delete without --prefix argument, that is actually what you want when you use one of the three filter arguments. the 'filter code' wouldn't be invoked is only invoked when at least one of these is used.
atm i can't think of an undesired outcome. do you have an example?

File: src/borg/testsuite/archiver.py (hunk 11 above)
Comment (enkore) [inline src/borg/testsuite/archiver.py]: How embarrassing :) I somehow thought this was main()

**Comment by funkyfuture** [comment]
ai'ght, thanks for the feedback. i changed the commits where the solution was clear to me.

File: src/borg/archiver.py (hunk 7 above)
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: it's seven instead of three.

it's a little less magic, your one-liner relies on short-circuit evaluation.

anyway, take what you like, it was just an option.

File: src/borg/helpers.py (hunk 9 above)
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: i meant if you change something, do the change in the commit where it belongs.
this can be done by first doing a small fixup commit and later merging that small fixup commit into the commit where the to-be-fixed code was introduced.

File: src/borg/archiver.py (hunk 12 above)
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: right, it is somehow consistency (how a validator name looks) vs. pep8 (different style for funcs and classes).

File: src/borg/archiver.py [hunk 16]
```diff
@@ -2281,6 +2368,23 @@ def build_parser(self, prog=None):
...
+        sort_by_default = 'timestamp'
+        filters_group.add_argument('--sort-by', dest='sort_by', type=sort_by_spec, default=sort_by_default,
+                               help='Comma-separated list of sorting keys; valid keys are: {}; default is: {}'
+                               .format(', '.join(HUMAN_SORT_KEYS), sort_by_default))
+
+        group = filters_group.add_mutually_exclusive_group()
+        group.add_argument('--first', dest='first', metavar='N', default=0, type=int,
+                           help='select first N archives')
+        group.add_argument('--last', dest='last', metavar='N', default=0, type=int,
+                           help='delete last N archives')
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: you call this method from list, delete, check, ... right?
so the help texts are still wrong.
i suggested the wording "consider first/last N archives" in a previous review.

File: src/borg/archiver.py [hunk 17]
```diff
@@ -2281,6 +2368,23 @@ def build_parser(self, prog=None):
                                help='hex object ID(s) to delete from the repo')
         return parser
 
+    @staticmethod
+    def add_archives_filters_args(subparser):
+        filters_group = subparser.add_argument_group('filters', 'Archive filters can be applied to repository targets.')
+        filters_group.add_argument('-P', '--prefix', dest='prefix', type=prefix_spec, default='',
+                                   help='only consider archive names starting with this prefix')
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: what happens with `borg delete --prefix repo` ?

Comment (funkyfuture) [inline src/borg/archiver.py]: if `BORG_REPOSITORY` is set, it would filter all archives with the prefix `repo`, otherwise it would complain about a missing location arg.

Comment (ThomasWaldmann) [inline src/borg/archiver.py]: OK, that was a bad example. ;)

File: src/borg/archiver.py (hunk 16 above)
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: ok

**Comment by funkyfuture** [comment]
okay, i did a fix commit concerning the remaining issues. to rebase it is kind of a pita.

**Comment by ThomasWaldmann** [comment]
any idea how to review d61c778869b9f336cb301682b0e01cf5f9f61786 ?

**Comment by funkyfuture** [comment]
maybe with two screens, reading both revisions side-by-side rather than with a diff view

**Comment by ThomasWaldmann** [comment]
Sorry, I can't review https://github.com/borgbackup/borg/commit/d61c778869b9f336cb301682b0e01cf5f9f61786 .

It somehow looks better than the old code, but I can not see whether it still does the same thing and works correctly (except what is covered by the still succeeding unit tests). At the same time, this is critical code with the capability to either not repair your repo or make stuff even worse, if things go wrong.

So, anybody else wants to review this?

Or, remove it from this PR to get the rest adopted more quickly?

In general, I would advise doing less complex/big changesets and especially hold back from refactoring / cleaning up and introducing new features within the same changeset (or even PR) - it just messes up the limited diff view on github and makes things hard to review.

**Comment by funkyfuture** [comment]
since you changed the ArchiveChecker and i was really not keen to rebase onto that, i removed the commit in question.

concerning the review, i have no other idea than (more?!) tests and reading.

**Comment by enkore** [comment]
Can you rebase this on current master? Thanks!

File: src/borg/archiver.py [hunk 18]
```diff
@@ -2290,6 +2344,23 @@ def build_parser(self, prog=None):
...
+
+        sort_by_default = 'timestamp'
+        filters_group.add_argument('--sort-by', dest='sort_by', type=SortBySpec, default=sort_by_default,
+                                   help='Comma-separated list of sorting keys; valid keys are: {}; default is: {}'
+                                   .format(', '.join(HUMAN_SORT_KEYS), sort_by_default))
+
+        group = filters_group.add_mutually_exclusive_group()
+        group.add_argument('--first', dest='first', metavar='N', default=0, type=int,
+                           help='consider first N archives after other filter args were applied')
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: nitpick: "after other filters were applied"

File: src/borg/archiver.py [hunk 19]
```diff
@@ -2290,6 +2344,23 @@ def build_parser(self, prog=None):
...
+        group.add_argument('--last', dest='last', metavar='N', default=0, type=int,
+                           help='consider last N archives after other filter args were applied')
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: same here

File: src/borg/helpers.py [hunk 20]
```diff
@@ -142,11 +142,15 @@ def __delitem__(self, name):
         name = safe_encode(name)
         del self._archives[name]
 
-    def list(self, sort_by=None, reverse=False):
-        # inexpensive Archive.list_archives replacement if we just need .name, .id, .ts
-        archives = self.values()  # [self[name] for name in self]
+    def list(self, sort_by=None, reverse=False, prefix=''):
+        """ Inexpensive Archive.list_archives replacement if we just need .name, .id, .ts
+            Returns list of borg.helpers.ArchiveInfo instances
+        """
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: docstring style:
"""
like
this
"""
or
"""like this"""

File: src/borg/archiver.py [hunk 21]
```diff
@@ -2352,6 +2423,25 @@ def run(self, args):
...
 
+    def _get_filtered_archives(self, args, manifest):
+        if args.location.archive:
+            raise Error('The options --first, --last and --prefix can only be used on repository targets.')
+
+        archives = manifest.archives.list(prefix=args.prefix)
+        if not archives:
+            logger.critical('There are no archives.')
+            self.exit_code = self.exit_code or EXIT_WARNING
+            return []
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: please remove the complete if-block, it is rather harmful than doing anything good.
if archives is empty here, the rest of the code here will still work normally and return a empty list at the end, no special casing needed.

**Comment by ThomasWaldmann** [comment]
@funkyfuture if we can finish this quickly, it can go into 1.1.0b2 still, release is tomorrow.

**Comment by funkyfuture** [comment]
i'll try to get it done this night.

**Comment by ThomasWaldmann** [comment]
Cool, thanks.

**Comment by funkyfuture** [comment]
can you give me a hint what is suddenly causing this:

```
user@host:~/python/borg (archives_filters) $ borg init --encryption=none /tmp/test
self test test_iteritems (borg.testsuite.hashindex.HashIndexTestCase) FAILED:
Traceback (most recent call last):
  File "/home/user/python/borg/src/borg/testsuite/hashindex.py", line 91, in test_iteritems
    self.assert_raises(StopIteration, next, iterator)
  File "/usr/lib/python3.5/site-packages/_pytest/python.py", line 1202, in raises
    pytest.fail(message)
  File "/usr/lib/python3.5/site-packages/_pytest/runner.py", line 537, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: DID NOT RAISE <class 'StopIteration'>

self test failed
This is a bug either in Borg or in the package / distribution you use.
```

**Comment by enkore** [comment]
Recompile the extensions (`python setup.py develop` or `pip install -e .`); it seems we forgot to bumb the API_VERSION there.

**Comment by funkyfuture** [comment]
i incorporated the demanded changes, but did not squash these as the effort and risk to break things are too high.

**Comment by ThomasWaldmann** [comment]
./src/borg/testsuite/archiver.py:1830:5: E303 too many blank lines (2)

**Comment by ThomasWaldmann** [comment]
lgtm modulo test results.

**Comment by funkyfuture** [comment]
got it.

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/issues/1663
